katsdpcontroller/product_config.py

```diff
@@ -31,63 +82,1196 @@ class _Sensor(ABC):
 }
 
 
+class Configuration:
+    def __init__(self, *,
+                 options: Options,
+                 simulation: Simulation,
+                 streams: Iterable[Stream]) -> None:
+        self.options = options
+        self.simulation = simulation
+        self.streams = list(streams)
+        self._by_class: Dict[Type[Stream], List[Stream]] = {}
+        for stream in streams:
+            self._by_class.setdefault(type(stream), []).append(stream)
+
+    def by_class(self, stream_cls: Type[_S]) -> Sequence[_S]:
+        return self._by_class.get(stream_cls, [])    # type: ignore
+
+    @classmethod
+    async def from_config(cls, config: Mapping[str, Any]) -> 'Configuration':
+        _validate(config)
+        config = _upgrade(config)
+        options = Options.from_config(config.get('config', {}))
+        simulation = Simulation.from_config(config.get('simulation', {}))
+        # First get the cam.http stream, so that sensors can be extracted
+        cam_http: Optional[CamHttpStream] = None
+        stream_configs = {**config.get('inputs', {}), **config.get('outputs', {})}
+        for name, stream_config in stream_configs.items():
+            if stream_config['type'] == 'cam.http':
+                cam_http = CamHttpStream.from_config(options, name, stream_config, [], {})
+                break
+
+        # Extract sensor values. This is pulled into a separate block rather
+        # than done as we construct each stream so that one day it can be
+        # optimised to make a single sensor_values call for all sensors.
+        sensors: Dict[str, Dict[str, Any]] = {name: {} for name in stream_configs}
+        if cam_http:
+            client = katportalclient.KATPortalClient(str(cam_http.url), None)
+            components = {}
+            for name in ['cbf', 'sub']:
+                try:
+                    components[name] = await client.sensor_subarray_lookup(name, None)
+                except Exception as exc:
+                    # There are too many possible exceptions from katportalclient to
+                    # try to list them all explicitly.
+                    raise SensorFailure(f'Could not get component name for {name}: {exc}') from exc
+            for name, stream_config in stream_configs.items():
+                stream_cls = STREAM_CLASSES[stream_config['type']]
+                for sensor in stream_cls._class_sensors:
+                    instrument = stream_config['instrument_dev_name']
+                    full_name = sensor.full_name(components, name, instrument)
+                    try:
+                        sample = await client.sensor_value(full_name)
+                    except Exception as exc:
+                        raise SensorFailure(f'Could not get value for {full_name}: {exc}') from exc
+                    if sample.status not in {'nominal', 'warn', 'error'}:
+                        raise SensorFailure(f'Sensor {full_name} has status {sample.status}')
```

Comment:
very minor: "...has invalid status..."

Reply:
I'm not sure I'd call other statuses "invalid": it's there are other statuses that are valid in katcp, they just indicate that the value isn't usable.

Reply:
Fine - I just felt that 'Sensor foo has status happy' doesn't convey our dismay at the status being unexpected. In fact, how about "...has unexpected status..."

Reply:
I'm happier with that. Done.